### DISCUSSION TITLE
Can we tune hyperparameters?

### DISCUSSION THREAD

**Discussion by Meelisha** [neuralhydrology/neuralhydrology#85] [Q&A] [answered]
Hello!

I was wondering if it was possible to tune the hyperparameters like epochs, batch-size, learning rate, etc. probably using sklearn or any other ML python packages that are out there to search for optimum parameters?

Thanks.

**Comment by kratzert** [comment, marked as answer]
Hi @Meelisha 

no automated option for hyper parameter tuning is currently implemented/supported from our side. Usually, we do grid search and use https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/utils/configutils.py#L9 to create configs for all possible parameter combinations and then use https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/nh_run_scheduler.py to run all configs across a GPU server managed by one job. If you have ideas for integrating a more automated process, happy to hear your ideas.

**Reply by Meelisha** [reply]
Hi @kratzert ,

Thank you for your response. 

I was thinking of doing a grid search as well but wasn't sure how I should be doing that with NeuralHydrology. 
Just a quick follow up question on that, how long did it take to run all the configs using the scheduler?

**Reply by kratzert** [reply]
Is it clear now how to do grid search, efficiently? Regarding runtime: all the run scheduler does is to take a bunch of configs and schedule jobs on different GPUs (you can usually run multiple training jobs on one GPU) and once a job is finished it starts the next job for the next config. So runtime is not really dependent on the scheduler. You can estimate the total runtime by extrapolating for the runtime of one config. Note that 2 jobs on one GPU usually make the runtime of an individual job usually a bit slower but overall is a net win. And more general: runtime depends on the number of basins/data lengthy hyper parameters etc. Really difficult to give you are more explicit answer here.

**Reply by Meelisha** [reply]
Well... I am not certain yet, but, what I understood is, you create all possible configurations and train the model (scheduler task) using each one of them and then choose the best combination based on the training/validation score. Forgive me if I missed something or got it completely off. 
I understand what you are saying about the variability of runtime being dependent on the input and hyperparameter. Thank you for the detailed explanation. I have not tried this approach yet but I would definitely give it a try soon.

**Reply by kratzert** [reply]
Exactly, what you described yes. The tools I linked are just utilities to make things easier (e.g. create configs for all possible hyper parameter combinations you want to test and to train models for all configs from one call).

**Reply by Meelisha** [reply]
Great!!! Thank you very much!
